mdbook-tera-backend/src/tera_renderer/renderer.rs

```diff
@@ -0,0 +1,152 @@
+    serialized_ctx: serde_json::Value,
+    components: Vec<CustomComponent>,
+    counter: u64,
+    tera_template: Tera,
+}
+
+impl Renderer {
+    pub(crate) fn new(ctx: RenderContext, tera_template: Tera) -> Result<Renderer> {
+        let mut renderer = Renderer {
+            serialized_ctx: serde_json::to_value(&ctx)?,
+            ctx: Arc::new(ctx),
+            components: Vec::new(),
+            counter: 0,
+            tera_template,
+        };
+        renderer
+            .tera_template
+            .register_function("get_context", renderer.create_get_context_function());
+        Ok(renderer)
+    }
+
+    pub(crate) fn add_component(&mut self, mut component: CustomComponent) {
+        component.register_function("get_context", self.create_get_context_function());
+        self.components.push(component);
+    }
+
+    fn create_get_context_function(&self) -> impl tera::Function {
+        let ctx_rc = Arc::clone(&self.ctx);
+        move |args: &HashMap<String, serde_json::value::Value>| -> tera::Result<tera::Value> {
+            let key = args
+                .get("key")
+                .ok_or_else(|| tera::Error::from(format!("No key argument provided")))?
+                .as_str()
+                .ok_or_else(|| {
+                    tera::Error::from(format!("Key has invalid type, expected string"))
+                })?;
+            let value = ctx_rc
+                .config
+                .get(key)
+                .ok_or_else(|| tera::Error::from(format!("Could not find key {key} in config")))?;
+            let value = to_value(value)?;
+            Ok(value)
+        }
+    }
+
+    pub(crate) fn render_book(&mut self) -> Result<()> {
+        let dest_dir = &self
+            .ctx
+            .destination
+            .parent()
+            .ok_or_else(|| {
+                anyhow!(
+                    "Destination directory {:?} has no parent",
+                    self.ctx.destination
+                )
+            })?
+            .to_owned();
+        if !dest_dir.is_dir() {
+            return Err(anyhow!("{:?} is not a directory", dest_dir));
+        }
+        self.render_book_directory(&dest_dir)
+    }
+
+    fn render_book_directory(&mut self, path: &Path) -> Result<()> {
+        for entry in path.read_dir()? {
+            let entry = entry?;
+            let path = entry.path();
+            if path.is_dir() {
+                self.render_book_directory(&path)?;
+            } else {
+                self.process_file(&path)?;
+            }
+        }
+        Ok(())
+    }
+
+    fn process_file(&mut self, path: &Path) -> Result<()> {
+        if path.extension().unwrap_or_default() != "html" {
+            return Ok(());
+        }
+        let mut file_content = String::new();
+        {
+            let mut file = fs::File::open(path)?;
+            file.read_to_string(&mut file_content)?;
+        }
+
+        let output = self.render_components(&file_content, path)?;
```

Comment:
Please rip out the notion of a "custom component" for now.

Instead, require people to write pure Tera code in the HTML files which you process. That would be the MVP for this feature. I would like us to land the core feature first before we solve potential down-stream problems.